### ISSUE TITLE
Surface source (HDF5 or MCPL) are written only for the last batch

### ISSUE THREAD

**Issue by marquezj** [openmc-dev/openmc#2502]
Hi, with @nicriz we are generating MCPL files and found out that OpenMC only writes them for the last batch:

https://github.com/openmc-dev/openmc/blob/4cadcbcf029310ced0eb3e7f638259f600335238/src/simulation.cpp#L427

Is there a reason for that? I would suggest replacing it with:

```
  if (settings::surf_source_write &&
      simulation::current_batch > settings::n_inactive) 
```

Does it make sense also for fixed source calculations?

Tagging also: @ebknudsen

**Comment by paulromano** [comment]
With that change, it would try to write the `surface_source.[h5,mcpl]` file at every (active) batch. Once the surface source bank is full, it wouldn't make sense to continually write the same exact file every batch.

**Comment by marquezj** [comment]
Sorry, my mistake. Yes, it makes sense to write it only at the end. 

What I actually meant is that the surface crossings are banked only for the last batch. I think it is here:

https://github.com/openmc-dev/openmc/blob/4cadcbcf029310ced0eb3e7f638259f600335238/src/particle.cpp#L403

(and I hope this time I got it right)

**Comment by gridley** [comment]
I agree that this is not necessarily the ideal behavior. The best strategy that minimizes bias would be to sample a few from all of the active cycles. Maybe we could estimate how many particles cross the surface per cycle, then add at most that amount from each active cycle until it becomes full.

**Comment by paulromano** [comment]
Sounds too complicated to me -- I think the right approach is to simply start banking during active batches as suggested by @marquezj.

**Comment by ebknudsen** [comment]
Potentially daft idea and not sure how complicated it would be to write (I could have a go): one could perhaps flush the bank to file when and if it is full (appending as we go along) and then at the end. For active batches that is. That would create more cross-talk during parallel runs though.

**Comment by paulromano** [comment]
Closed with merging of #2592
